### ISSUE TITLE
Docker save operation for more images will fail when delete one of the choose images.

### ISSUE THREAD

**Issue by charlihappy2016** [moby/moby#29653]
Description
Version: docker-1.11.2

Steps to reproduce the issue:
1. Execute docker save for more than one images, 
2. During that process delete one of the choose images but not the first one. 

The result of the save operation will fail with the prompt like below:
 Error response from daemon: open /var/lib/docker/image/devicemapper/imagedb/content/sha256/5d360e2f0ea02af74a86d23a779fe85c2f81b6d7d3b958ced665760be63b139d: no such file or directory

From the prompt info, the reason is when save operation needs the images, but the delete is done, so the images is not there. 
Maybe in the save process the images should be locked first.

**Comment by thaJeztah** [comment]
Can you provide some more information?

- first of all, please include the output of `docker version` and `docker info`, as requested in the [issue report template](https://raw.githubusercontent.com/docker/docker/master/.github/ISSUE_TEMPLATE.md) when opening your issue
- when you mention "but not the first one", you mean not the image that the container was started from, but one of its "parent images"?
- did you use `-f` / `--force` when removing images?
- ideally, provide the _exact_ commands to run (and how / in what order), or a small script to show how to reproduce the issue

**Comment by charlihappy2016** [comment]
	Docker version and info
# Docker version
Client:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:30:42 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.5
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   7392c3b
 Built:        Fri Dec 16 02:30:42 2016
 OS/Arch:      linux/amd64

# Docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 1.12.5
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 73
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay null host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.13.0-24-generic
Operating System: Ubuntu 14.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.861 GiB
Name: ubuntu
ID: XC5Y:RVTS:DCNV:MHBG:47SE:QPHU:GMJS:AYIQ:6526:IUNQ:YUDW:RNOT
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
WARNING: No kernel memory limit support
 127.0.0.0/8

	The first one means the images used to do save operation. 
	Execute commands:
1. Do image save operation for two images: ubuntu  and  busybox
docker save –o test.tar ubuntu busybox
2. During that process delete the images : busybox
docker rmi busybox
3. the save result is fail with the log;
root@ubuntu:~# docker save -o test.tar ubuntu busybox
Error response from daemon: open /var/lib/docker/image/aufs/imagedb/content/sha256/e4415b714b624040f19f45994b51daed5cbdb00e0eb9a07221ff0bd6bcf55ed7: no such file or directory

if use -f/--force to remove the images:
docker rmi -f busybox
the result is:
root@ubuntu:~# docker save -o test.tar ubuntu busybox
Error response from daemon: layer does not exist

In a word, the save result is fail if remove the images.

**Comment by moypray** [comment]
Hi, @thaJeztah I met the same issue too. And I have found the root cause of this issue: 
```
func (l *tarexporter) Save(names []string, outStream io.Writer) error {
        // Here the function parseNames will get all the images which given by user.
        images, err := l.parseNames(names)  
        if err != nil {
                return err
        }

         // And the save function will use l.is.Get() to get the images to be saved.
         // While after get the image info and before l.is.Get is called, if we remove this image, will return the error which we met.
        return (&saveSession{tarexporter: l, images: images}).save(outStream)
}
```

I think we should add a lightweight `imageStateMgr`, if we save a image, we set the image to savingInProgress state, and if we want to delete this image, we should check if the image is in some state like 'savingInProcess' bla bla. Only the images which are in some special operation are stored here.  
And of course other image operation dependence could be handled here too.  
Correct me if something is wrong.

**Comment by sam-thibault** [comment]
fixed by #33323
